### PR TITLE
Feathr UI: API spec alignment and ux experience improvement

### PR DESCRIPTION
**Pull request by blrchen**
This PR update ui code to align with latest api spec which abstracts common interfaces among different registration providers. This PR also enhances lineage ux experience and adds RBAC experience for flighting.

Staging env: https://brave-moss-045d2ce00-6.eastasia.1.azurestaticapps.net/

## New Feature registration API integration
API spec can be viewed at https://feathr-registry.azurewebsites.net/docs.  UI code is now purview free.

## UX improvements
- [x] When a feature is clicked, the feature node is highlighted
- [x] When a feature is clicked, the feature level lineage path is highlighted. Feature level lineage means direct producer and consumer.
- [x] Node in graph are colored grouped by feature type.
- [x] User can toggle graph to filter by feature type.
- [x] Flow arrow is enabled to indicate producer and consumer relationship
- [x] All out of scope resources are removed in graph, for example, Altas Process, Project.
- [x] Add navigate button in node so user can click to navigate to feature detail page.
- [x] RBAC in flgihting, accessible via /management path. 

Linked issue: #272

### CODE REVIEW

**Comment by hangfei** [comment]
Overall it's pretty good. Some minor comments:

1. After i clicked a certain figure, can we highlight the feature where i clicked?
2. slightly move the graph in the middle.
3. Add tooltip or explantion text for what each color means
4. Add a way to click the feature and open the detail page.

File: ui/package.json [hunk 1]
```diff
@@ -10,11 +10,9 @@
     "dagre": "^0.8.5",
     "react": "^17.0.2",
     "react-dom": "^17.0.2",
-    "react-flow-renderer": "^10.2.1",
+    "react-flow-renderer": "^9.7.4",
```
Comment (hangfei) [inline ui/package.json:13]: why the downgrade?

File: ui/src/api/api.tsx [hunk 2]
```diff
@@ -46,6 +46,14 @@ export const fetchFeature = async (project: string, qualifiedName: string) => {
 };
 
 export const fetchProjectLineages = async (project: string) => {
+  return Axios
+    .get<IFeatureLineage>(`${ API_ENDPOINT }/v1/purview/${ purview }/projects/${ project }?code=${ token }`, {})
```
Comment (hangfei) [inline ui/src/api/api.tsx:50]: it says SQL in API endpoint. But here it's still related purview. Can you fix it?

File: ui/src/app.tsx [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+const queryClient = new QueryClient();
+
+const msalConfig: Configuration = {
+  auth: {
+    clientId: process.env.REACT_APP_AAD_APP_CLIENT_ID,
+    authority: process.env.REACT_APP_AAD_APP_AUTHORITY,
+    redirectUri: window.location.origin,
+  },
+};
+const pca = new PublicClientApplication(msalConfig);
```
Comment (hangfei) [inline ui/src/app.tsx:27]: avoid acronym

File: ui/src/app.tsx [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import SideMenu from "./components/sidemenu/siteMenu";
+import Features from "./pages/feature/features";
+import NewFeature from "./pages/feature/newFeature";
+import FeatureDetails from "./pages/feature/featureDetails";
+import DataSources from "./pages/dataSource/dataSources";
+import Jobs from "./pages/jobs/jobs";
+import Monitoring from "./pages/monitoring/monitoring";
+import LineageGraph from "./pages/feature/lineageGraph";
+
+type Props = {};
```
Comment (hangfei) [inline ui/src/app.tsx:17]: the Props are empty. Do we still need it?

File: ui/src/components/lineage/graph.tsx [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+  }, [data, nodeId]);
+
+  const resetHighlight = (): void => {
+    if (!elements || elements.length === 0) {
+      return;
+    }
+
+    const values: Elements = [];
+
+    for (let index = 0; index < elements.length; index++) {
```
Comment (hangfei) [inline ui/src/components/lineage/graph.tsx:47]: will x in elements syntax be better?

File: ui/src/components/lineage/graph.tsx [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+          ...element,
+          animated: false,
+        });
+      }
+    }
+
+    setElements(values);
+  };
+
+  const highlightPath = (node: Node, check: boolean): void => {
```
Comment (hangfei) [inline ui/src/components/lineage/graph.tsx:70]: add dev doc

File: ui/src/components/lineage/graph.tsx [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+          ...element,
+          animated: highlight,
+        });
+      }
+    }
+
+    setElements(values);
+  };
+
+  const fitElements = (): void => {
```
Comment (hangfei) [inline ui/src/components/lineage/graph.tsx:116]: add dev doc

File: ui/src/components/lineage/graphNode.tsx [hunk 8]
```diff
@@ -0,0 +1,45 @@
+import React, { FC, memo } from 'react';
+import { RightCircleOutlined } from "@ant-design/icons";
+import { Handle, NodeProps, Position } from 'react-flow-renderer';
+import { useNavigate, useParams } from "react-router-dom";
+
+type Params = {
+  project: string;
+}
+const GraphNode: FC<NodeProps> = (props: NodeProps) => {
```
Comment (hangfei) [inline ui/src/components/lineage/graphNode.tsx:9]: add doc

File: ui/src/pages/feature/lineageGraph.tsx [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+  return (
+    <Card title={ "Lineage" }>
+      <div>
+        <Radio.Group onChange={ e => toggleFeatureType(e.target.value) }>
+          <Radio.Button value="feathr_source_v1"> Source </Radio.Button>
+          <Radio.Button value="feathr_anchor_v1">Anchor</Radio.Button>
+          <Radio.Button value="feathr_anchor_feature_v1">Anchor Feature</Radio.Button>
+          <Radio.Button value="feathr_derived_feature_v1">Derived Feature</Radio.Button>
+          <Radio.Button value="all_nodes">All Features</Radio.Button>
```
Comment (hangfei) [inline ui/src/pages/feature/lineageGraph.tsx:91]: put all features as firste element. Also highlight all features. Right now, it's grey upon loading.

File: ui/src/pages/feature/lineageGraph.tsx [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+import { useParams, useSearchParams } from "react-router-dom";
+import { Elements } from 'react-flow-renderer';
+import Graph from "../../components/lineage/graph";
+import { generateEdge, generateNode } from "../../components/lineage/utils";
+import { fetchProjectLineages } from "../../api";
+
+type Params = {
+  project: string;
+}
+const LineageGraph: React.FC = () => {
```
Comment (hangfei) [inline ui/src/pages/feature/lineageGraph.tsx:12]: add docs

File: ui/src/components/lineage/utils.ts [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+  feathr_source_v1: 'hsl(315, 100%, 50%)',
+  feathr_anchor_v1: 'hsl(270, 100%, 50%)',
+  feathr_anchor_feature_v1: 'hsl(225, 100%, 50%)',
+  feathr_derived_feature_v1: 'hsl(135, 100%, 50%)'
+};
+
+const generateNode = ({
+                        nodeId, index,
+                        currentNode
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Comment (hangfei) [inline ui/src/components/lineage/utils.ts:72]: do we still need this?

File: ui/src/api/api.tsx (hunk 2 above)
Comment (blrchen) [reply ui/src/api/api.tsx:50]: This is intended right now. SQL based registration api is compatible with Purview based registration api. The purview account name will be passed but not used in SQL implementation. 

In a further version. Purview based registration api will remove purview accounts as parameter and move it to server side config. Then all purview related config will be removed in ui code as well.

File: ui/package.json (hunk 1 above)
Comment (blrchen) [reply ui/package.json:13]: Some code works in v9 have issues in v10. Indented to use v9 for now as v9 is more popular, download volume v9 latest is 60k while v10 is a few hundreds.

File: ui/src/pages/feature/lineageGraph.tsx (hunk 9 above)
Comment (blrchen) [reply ui/src/pages/feature/lineageGraph.tsx:91]: Good suggestion. Moved All Features toggle button to the first position and make it active by default.

File: ui/src/components/lineage/utils.ts (hunk 11 above)
Comment (blrchen) [reply ui/src/components/lineage/utils.ts]: It's recommended to not to use any but since api contract would be changed soon to decoupled with purview. Let me address this in a following up PR when new API is out

File: ui/src/api/api.tsx [hunk 12]
```diff
@@ -101,3 +102,41 @@ export const deleteFeature = async (qualifiedName: string) => {
...
+  .post(`${ API_ENDPOINT }/user/${roleForm.userName}/userroles/delete`, roleForm,
+      {
+        headers: { "Content-Type": "application/json;" },
+        params: {},
+      }).then((response) => {
+      return response;
+    }).catch((error) => {
+      return error.response;
+    });
+}
```
Comment (hangfei) [inline ui/src/api/api.tsx:142]: it's not very good practice to bundle different things into one PR. Any special reason to do so? If not, let's do it in a separate PR. @Yuqing-cat

File: ui/src/api/api.tsx (hunk 2 above)
Comment (hangfei) [reply ui/src/api/api.tsx:51]: let's put a TODO here.

File: ui/src/api/api.tsx (hunk 12 above)
Comment (Yuqing-cat) [reply ui/src/api/api.tsx:142]: Hi @hangfei I understand your concerns. Let me explain the reason here:
The lineage re-design PR contains some upgrades and restructure in frontend code structure. It's impossible to split all RBAC frontend content into a separate PR based on current main branch. In another word, the re-design PR is a hard dependency for all frontend changes. 

I strongly suggest us including it in this time to save engineer effort.
(Btw, the RBAC page is hided right now, users will not be able to access unless they know the redirect URL.)

Comment (hangfei) [reply ui/src/api/api.tsx:142]: That works. Please update the PR description and the reasons to do so.

File: ui/src/models/model.ts [hunk 13]
```diff
@@ -74,3 +67,20 @@ export interface IFeatureLineage {
...
+  scope: string;
+  userName: string;
+  roleName: string;
+  createTime: string;
+  createReason: string;
+  deleteTime?: any;
+  deleteReason?: any;
+}
+
+export interface RoleForm {
```
Comment (hangfei) [inline ui/src/models/model.ts:81]: Though i am not a big fan of Ixxx. But we should follow the convention. So this prob should be IRoleForm

File: ui/src/models/model.ts [hunk 14]
```diff
@@ -1,35 +1,28 @@
...
   transformation: FeatureTransformation;
   key: FeatureKey[];
   window: string;
-  input_anchor_features: InputAnchorFeatures[];
-  input_derived_features: InputDerivedFeatures[]
+  _input_anchor_features: IFeature[];
+  _input_derived_features: IFeature[]
+}
+
+export interface FeatureType {
```
Comment (hangfei) [inline ui/src/models/model.ts:21]: IFeatureType

File: ui/src/models/model.ts [hunk 15]
```diff
@@ -1,35 +1,28 @@
...
+  type: string,
+  tensor_category: string,
+  dimension_type: string[],
+  val_type: string
 }
 
 export interface InputAnchorFeatures {
```
Comment (hangfei) [inline ui/src/models/model.ts:28]: IInputAnchorFeatures

**Comment by hangfei** [comment]
Let's not put two distinct commits into one PR in the future.

**Comment by hangfei** [comment]
I can't really find the UI changes for RBAC. If it's not accessible from UI, please attach a screenshot.

**Comment by blrchen** [comment]
> I can't really find the UI changes for RBAC. If it's not accessible from UI, please attach a screenshot.

RBAC UI is in flgihting, to pilot, plese use this direct url : https://brave-moss-045d2ce00-6.eastasia.1.azurestaticapps.net/management

File: ui/src/models/model.ts (hunk 13 above)
Comment (blrchen) [reply ui/src/models/model.ts:81]: I think we don't need to Ixxx here as there is not scenario in ui code requires different implementations. I will do a clean up to remove Ixxx pattern

**Review by hangfei — APPROVED** [review]

**Review by xiaoyongzhu — APPROVED** [review]
